File: frontend/src/api/wiki.ts
import api from './client'
import type { User } from './auth'

export type WikiVisibility = 'public' | 'restricted' | 'private'

export interface WikiSection {
  id: number
  title: string
  parent_id: number | null
}

export interface WikiArticleListItem {
  id: number
  title: string
  section_id: number | null
  tags: string[]
  visibility: WikiVisibility
  updated_at: string
  author: User
}

export interface WikiArticle extends WikiArticleListItem {
  content: string
  created_at: string
  section: WikiSection | null
  permissions: { user_id: number; user: User }[]
}

export interface WikiVersion {
  id: number
  article_id: number
  created_at: string
  author: User
}

export interface WikiArticleCreate {
  title: string
  content?: string
  section_id?: number | null
  tags?: string[]
  visibility?: WikiVisibility
  allowed_user_ids?: number[]
}

export interface WikiArticleUpdate {
  title?: string
  content?: string
  section_id?: number | null
  tags?: string[]
  visibility?: WikiVisibility
  allowed_user_ids?: number[]
}

export async function listSections(): Promise<WikiSection[]> {
  const { data } = await api.get<WikiSection[]>('/wiki/sections')
  return data
}

export async function createSection(title: string, parent_id?: number | null): Promise<WikiSection> {
  const { data } = await api.post<WikiSection>('/wiki/sections', { title, parent_id })
  return data
}

export async function listArticles(params?: {
  section_id?: number
  tag?: string
  skip?: number
  limit?: number
}): Promise<WikiArticleListItem[]> {
  const { data } = await api.get<WikiArticleListItem[]>('/wiki/articles', { params })
  return data
}

export async function getArticle(id: number): Promise<WikiArticle> {
  const { data } = await api.get<WikiArticle>(`/wiki/articles/${id}`)
  return data
}

export async function createArticle(body: WikiArticleCreate): Promise<WikiArticle> {
  const { data } = await api.post<WikiArticle>('/wiki/articles', body)
  return data
}

export async function updateArticle(id: number, body: WikiArticleUpdate): Promise<WikiArticle> {
  const { data } = await api.patch<WikiArticle>(`/wiki/articles/${id}`, body)
  return data
}

export async function deleteArticle(id: number): Promise<void> {
  await api.delete(`/wiki/articles/${id}`)
}

export async function listVersions(articleId: number): Promise<WikiVersion[]> {
  const { data } = await api.get<WikiVersion[]>(`/wiki/articles/${articleId}/versions`)
  return data
}

export async function restoreVersion(articleId: number, versionId: number): Promise<WikiArticle> {
  const { data } = await api.post<WikiArticle>(`/wiki/articles/${articleId}/restore/${versionId}`)
  return data
}

export async function uploadImage(file: File): Promise<string> {
  const form = new FormData()
  form.append('file', file)
  const { data } = await api.post<{ url: string }>('/wiki/images', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data.url
}

File: backend/schemas/wiki.py
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, Field

from models.wiki import WikiVisibility
from schemas.user import UserResponse


class WikiSectionCreate(BaseModel):
    title: str
    parent_id: int | None = None


class WikiSectionResponse(BaseModel):
    id: int
    title: str
    parent_id: int | None

    model_config = {"from_attributes": True}


class WikiArticleCreate(BaseModel):
    title: str
    content: str = ""
    section_id: int | None = None
    tags: list[str] = []
    visibility: WikiVisibility = WikiVisibility.public
    # allowed_user_ids используется только при visibility=restricted
    allowed_user_ids: list[Annotated[int, Field(gt=0)]] = []


class WikiArticleUpdate(BaseModel):
    title: str | None = None
    content: str | None = None
    section_id: int | None = None
    tags: list[str] | None = None
    visibility: WikiVisibility | None = None
    allowed_user_ids: list[Annotated[int, Field(gt=0)]] | None = None


class PermissionUserResponse(BaseModel):
    user_id: int
    user: UserResponse

    model_config = {"from_attributes": True}


class WikiArticleResponse(BaseModel):
    id: int
    title: str
    content: str
    section_id: int | None
    tags: list[str]
    visibility: WikiVisibility
    created_at: datetime
    updated_at: datetime
    author: UserResponse
    section: WikiSectionResponse | None
    permissions: list[PermissionUserResponse]

    model_config = {"from_attributes": True}


class WikiArticleListItem(BaseModel):
    """Краткая карточка для списка статей — без тяжёлого content."""

    id: int
    title: str
    section_id: int | None
    tags: list[str]
    visibility: WikiVisibility
    updated_at: datetime
    author: UserResponse

    model_config = {"from_attributes": True}


class WikiVersionResponse(BaseModel):
    id: int
    article_id: int
    created_at: datetime
    author: UserResponse

    model_config = {"from_attributes": True}
